lib/Index/IndexRecord.cpp

```diff
@@ -393,6 +396,195 @@ emitDataForSwiftSerializedModule(ModuleDecl *module,
+                            std::string &error) {
+  llvm::sys::fs::file_status unitStat;
+  if (std::error_code ec = llvm::sys::fs::status(filePath, unitStat)) {
+    if (ec != std::errc::no_such_file_or_directory) {
+      llvm::raw_string_ostream err(error);
+      err << "could not access path '" << filePath << "': " << ec.message();
+      return {};
+    }
+    return false;
+  }
+
+  if (!timeCompareFilePath)
+    return true;
+
+  llvm::sys::fs::file_status compareStat;
+  if (std::error_code ec =
+          llvm::sys::fs::status(*timeCompareFilePath, compareStat)) {
+    if (ec != std::errc::no_such_file_or_directory) {
+      llvm::raw_string_ostream err(error);
+      err << "could not access path '" << *timeCompareFilePath
+          << "': " << ec.message();
+      return {};
+    }
+    return true;
+  }
+
+  // Return true (unit is up-to-date) if the file to compare is older than the
+  // unit file.
+  return compareStat.getLastModificationTime() <=
+         unitStat.getLastModificationTime();
+}
+
+/// Emit the symbolic swift interface file for an imported Clang module into the
+/// index store directory.
+///
+/// The swift interface file is emitted only when it doesn't exist yet, or when
+/// the PCM for the Clang module has been updated.
+///
+/// System modules without the 'cplusplus' requirement are not emitted.
+static void emitSymbolicInterfaceForClangModule(
+    ClangModuleUnit *clangModUnit, ModuleDecl *M,
+    const clang::Module *clangModule, StringRef indexStorePath,
+    const clang::CompilerInstance &clangCI, DiagnosticEngine &diags) {
+  if (!M->getASTContext().LangOpts.EnableCXXInterop)
+    return;
+  // Skip system modules without an explicit 'cplusplus' requirement.
+  bool isSystem = clangModUnit->isSystemModule();
+  if (isSystem && !importer::requiresCPlusPlus(clangModule))
+    return;
+
+  // Make sure the `interfaces` directory is created.
+  std::string error;
+  if (initSymbolicInterfaceStorePath(indexStorePath, error)) {
+    diags.diagnose(SourceLoc(), diag::error_create_index_dir, error);
+    return;
+  }
+
+  // Determine the output name for the symbolic interface file.
+  clang::serialization::ModuleFile *ModFile =
+      clangCI.getASTReader()->getModuleManager().lookup(
+          clangModule->getASTFile());
+  assert(ModFile && "no module file loaded for module ?");
+  SmallString<128> interfaceOutputPath = indexStorePath;
+  appendSymbolicInterfaceToIndexStorePath(interfaceOutputPath);
+  appendSymbolicInterfaceClangModuleFilename(ModFile->FileName,
+                                             interfaceOutputPath);
+
+  // Check if the symbolic interface file is already up to date.
+  auto upToDate = isFileUpToDateForOutputFile(
+      interfaceOutputPath, StringRef(ModFile->FileName), error);
+  if (!upToDate) {
+    diags.diagnose(SourceLoc(), diag::error_index_failed_status_check, error);
+    return;
+  }
+  if (M->getASTContext().LangOpts.EnableIndexingSystemModuleRemarks) {
```

Comment:
Might be worth renaming to `EnableIndexingModuleRemarks`.

Reply:
I can do that as a follow-up.